### ISSUE TITLE
Upgrade apache-httpclient to httpclient5 which uses SLF4J

### ISSUE THREAD

**Issue by frehov** [aws/aws-sdk-java-v2#2831]
### Describe the feature

I'd like to request an upgrade of the apache httpclient to httpclient5 which uses SLF4J as a logging-facade instead of having to install bridge handlers to redirect log output from commons-logging.

### Is your Feature Request related to a problem?

It's not directly a problem, but the issue is that the apache httpclient is leaking a dependency on commons-logging which has to be bridged by either `jul-to-slf4j` or `jcl-over-slf4j` which is an inconvenience for downstream libraries.

This feature request is opened in relation to https://github.com/micronaut-projects/micronaut-aws/issues/1203

but it also seems it's a cause of frustration here as well https://github.com/quarkusio/quarkus/issues/4528

### Proposed Solution

My proposed solution would be to migrate from httpclient 4.5 til 5.1 following the migration guide set forward at the apache site https://hc.apache.org/httpcomponents-client-5.1.x/migration-guide/index.html

### Describe alternatives you've considered

There is a possbility for any downstream library that use dependencies from the aws-sdk-v2 to include bridge handlers to redirect the logs from the httpclient, but imho the solution should be done at the root and trickle down, simplifying library maintainers jobs by not having to actively exclude the commons-logging and installing bridge handlers.

### Acknowledge

- [ ] I may be able to implement this feature request

### AWS Java SDK version used

2.16.104 and later

### JDK version used

11

### Operating System and version

Linux

**Comment by debora-ito** [comment]
@frehov apologies for the long silence here. Marking this for review from the team.

**Comment by debora-ito** [comment]
Hi @frehov thank you for your patience.

Discussed it with the team, and apparently we tried to upgrade before but httpclient5 introduced breaking changes, and the work needed to make the changes on the SDK was not prioritized when compared to other tasks. We currently have no plans to make the upgrade.

Let us know if you have more questions.

**Comment by debora-ito** [comment]
@lazystone I'll raise this for discussion with the team, as it looks like [httpcore 4 is at EOL](https://hc.apache.org/httpcomponents-core-4.4.x/index.html) - httpclient 4 will probably follow the same path. Will post an update when we get to a conclusion.

**Comment by joebancf** [comment]
I'd just like to add some weight for the team discussion. httpcore v4 doesn't work with virtual threads and there doesn't seem to be a plan to backport the fix to v4. Therefore the AWS SDK is not compatible with virtual threads until the switch to v5 is made.

To be clear, virtual threads do technically work but they are pinned to their carrier threads due to the use of the synchronized keyword in httpcore. This makes them practically unusable in production.

**Comment by PankajSAgarwal** [comment]
Is there any plan to support aws sdk2 java with httpclient5 @debora-ito ?

**Comment by debora-ito** [comment]
We are considering migrating to httpclient5 and httpcore5, since 4 is at EOL. We need to do research on how to make the change in a backwards compatible way, and how to prioritize this task against other tasks we need to work on. Added a task in our backlog, we don't have a timeline to share right now.

**Comment by lazystone** [comment]
If this ticket is still on your backlog, then I suggest to prioritize #1447 before this one. #1447 solves future compatibility and Virtual Threads usage at the same time.

**Comment by rhermes62** [comment]
Our team just had a outage due to not knowing the AWS SDK's underlying HTTP client was incompatible with virtual threads. Pre-production stages did not have enough traffic to simulate the errors that result from issues like [this](https://issues.apache.org/jira/projects/HTTPCORE/issues/HTTPCORE-746?filter=allissues) until it was unfortunately too late.

The AWS SDK not currently supporting virtual threads should be more prominently documented until addressed to try and spare others from making this mistake while trying to develop modern Java applications.

**Comment by millems** [comment]
@rhermes62 Sorry this bit you. We're tracking an action item to document our state of support for virtual threads without pinning. We also have a desire to support HTTP clients that do not have thread pinning issues at some point in the future.

**Comment by patrickjamesbarry** [comment]
I agree with raising the transparency of client "pitfalls"  in documentation and the fact aws uses an unsupported version of the apache client.  https://docs.aws.amazon.com/sdk-for-java/latest/developer-guide/http-configuration.html would be the perfect spot to mention the usage of apache client 4 vs 5 and expected release date of client5 support.

**Comment by nitulkukadia** [comment]
This issue is open since Nov 9, 2021 , is the there any plan to upgrade to apache client 5.x ?

**Comment by Sakelig** [comment]
Even AWS Inspector is catching on to this with several vulernabilities, please update asap

**Comment by Sakelig** [comment]
please upgrade 🙈

**Comment by uros-ogrizovic** [comment]
@debora-ito any news on this?

**Comment by Javi3Code** [comment]
We have had a lot of deadlocks in the projects I collaborate in because of this issue, until someone found out that the aws sdk was not compatible with vthreads.

Numerous events lost and hours of support added.

Seriously... apache uploaded in July the upgrade, will we have it by 2030? No problem, then Oracle will have done its job...

**Comment by bekoenig** [comment]
Hej @debora-ito,

our framework stack for our company supplies a builder for all implementations and libraries which uses http client. This is useful for an easy configuration and implementation of customizers. It's bad to hear, that aws-sdk still uses http client 4 and has no support for http client 5. So we have to adapt the missing support or repeat the configuration pattern for http client 4.

Maybe aws-sdk can benefit of the [openrewrite recipes](https://docs.openrewrite.org/recipes/apache/httpclient5) too ;-).

Greetings,
Ben

**Comment by ghilainm** [comment]
Any news on this? Also blocked by the virtual thread pinning issue.

**Comment by debora-ito** [comment]
Quick update: yes we are working on the httpclient5 upgrade, it's on early stages of development, no timeline to share.

**Comment by nitulkukadia** [comment]
Hi @debora-ito  Please let me know how can I contribute.

**Comment by SoerenHenning** [comment]
Thanks for the update, @debora-ito. If upgrading the Apache HttpClient to version 5 is such a big effort, could you share why it is prioritized over implementing a client based on the HTTP client that comes with JDK since Java 11 (see also https://github.com/aws/aws-sdk-java-v2/issues/1447)?

**Comment by debora-ito** [comment]
@SoerenHenning The main point is Apache HttpClient is the default sync client in the SDK, we need to upgrade to 5.x since 4.x is deprecated. We've already needed some bugfixes that were released in Apache HttpClient 5.x but were not backported to 4.x.

**Comment by SoerenHenning** [comment]
Hi @debora-ito, it looks like there is a lot of activity on https://github.com/aws/aws-sdk-java-v2/tree/feature/master/apache5x. Is there any public information for when a release is planned or how far the integration is already?

**Comment by debora-ito** [comment]
Hey all,

We're excited to announce the Developer Preview of Apache 5 HTTP Client! Check the [announcement](https://github.com/aws/aws-sdk-java-v2/issues/6268) for more info, try it out today and send us your feedback.

**Comment by marcreichman-pfi** [comment]
@debora-ito do you have a rough idea when we'd expect transitive dependencies for things like s3/s3 transfer manager to be based on the apache5 setup? Even if one were to switch their own declarative dependency to the preview apache5 client, for example, to use the client builder and tweak settings, the existing transitive dependency will still bring in httpclient 4.x artifacts.

**Comment by ziqin** [comment]
@marcreichman-pfi

> Even if one were to switch their own declarative dependency to the preview apache5 client, for example, to use the client builder and tweak settings, the existing transitive dependency will still bring in httpclient 4.x artifacts.

You can simplify exclude the transitive dependency with your dependency manager.

If you are using Spring Cloud for AWS with gradle like me, you can use the following configuration:

```groovy
dependencies {
  implementation platform("software.amazon.awssdk:bom:${awsJavaSdkVersion}")
  implementation platform("io.awspring.cloud:spring-cloud-aws-dependencies:${springCloudAwsVersion}")
  implementation('io.awspring.cloud:spring-cloud-aws-starter-s3') {
    exclude group: 'software.amazon.awssdk', module: 'apache-client'
  }
  implementation 'software.amazon.awssdk:apache5-client'
}
```

or, with the `io.spring.dependency-management` plugin applied:

```groovy
dependencies {
  implementation('io.awspring.cloud:spring-cloud-aws-starter-s3') {
    exclude group: 'software.amazon.awssdk', module: 'apache-client'
  }
  implementation 'software.amazon.awssdk:apache5-client'
}
dependencyManagement {
  imports {
    mavenBom "io.awspring.cloud:spring-cloud-aws-dependencies:${springCloudAwsVersion}"
    mavenBom "software.amazon.awssdk:bom:${awsJavaSdkVersion}"
  }
}
```

**Comment by marcreichman-pfi** [comment]
@ziqin Thanks for that tip! When the world trusts this change eventually I'll try that out.

**Comment by dagnir** [comment]
Hi all! I'm excited to announce that the Apache5 HTTP Client is now generally available! Please see our [announcement post for more information](https://github.com/aws/aws-sdk-java-v2/issues/6657).
